crates/predicate/src/verifiers/sp1_groth16.rs

```diff
@@ -0,0 +1,210 @@
+/// 5. Returns success if either hash method validates the proof
+///
+/// SP1's Groth16 circuit expects two public inputs:
+/// - `program_id`: Identifier of the SP1 program (embedded in verifying key)
+/// - `hash(public_values)`: SHA-256 or Blake3 hash of the claim data
+#[derive(Debug, Default)]
+pub(crate) struct Sp1Groth16Verifier;
+
+impl PredicateVerifier for Sp1Groth16Verifier {
+    type Condition = Groth16VerifyingKey;
+    type Witness = Groth16Proof;
+
+    fn parse_condition(&self, condition: &[u8]) -> PredicateResult<Self::Condition> {
+        // Parse condition bytes as SP1Groth16Verifier (contains program ID + verifying key)
+        borsh::from_slice(condition).map_err(|e| PredicateError::PredicateParsingFailed {
+            id: PredicateTypeId::Sp1Groth16,
+            reason: format!("failed to parse SP1 Groth16 verifying key from condition data: {e}",),
+        })
+    }
+
+    fn parse_witness(&self, witness: &[u8]) -> PredicateResult<Self::Witness> {
+        // Parse witness bytes as Groth16Proof using zkaleido's loader
+        Groth16Proof::load_from_gnark_bytes(witness).map_err(|e| {
+            PredicateError::WitnessParsingFailed {
+                id: PredicateTypeId::Sp1Groth16,
+                reason: format!("failed to parse Groth16 proof from witness data: {e}"),
+            }
+        })
+    }
+
+    fn verify_inner(
+        &self,
+        program: &Self::Condition,
+        claim: &[u8],
+        proof: &Self::Witness,
+    ) -> PredicateResult<()> {
+        // SP1's Groth16 circuit expects two public inputs:
+        // 1. program_id (embedded in the verifying key)
+        // 2. hash(public_values) (computed from claim)
+
+        // Try SHA-256 hash first as it's the default for SP1
+        let fr_sha2 = sha256_to_fr(claim).map_err(|e| PredicateError::VerificationFailed {
+            id: PredicateTypeId::Sp1Groth16,
+            reason: format!("failed to compute SHA-256 hash of claim: {e}"),
+        })?;
+
+        // Attempt algebraic verification using zkaleido's verifier with SHA-256 hash
+        if verify_sp1_groth16_algebraic(program, proof, &fr_sha2).is_ok() {
+            return Ok(());
+        }
+
+        // Fallback: try Blake3 hash for compatibility with different SP1 configurations
+        let fr_blake3 = blake3_to_fr(claim).map_err(|e| PredicateError::VerificationFailed {
+            id: PredicateTypeId::Sp1Groth16,
+            reason: format!("failed to compute Blake3 hash of claim: {e}"),
+        })?;
+
+        // Retry verification with Blake3 hash using zkaleido's verifier
+        verify_sp1_groth16_algebraic(program, proof, &fr_blake3).map_err(|e| {
+            PredicateError::VerificationFailed {
+                id: PredicateTypeId::Sp1Groth16,
+                reason: format!("SP1 Groth16 proof verification failed with both SHA-256 and Blake3 hashing: {e}"),
+            }
+        })
```

Comment:
This feels sketchy to me. Is there a case where we'd reasonably want this behavior, and where the caller wouldn't know which hash to expect? Seems safer to make the two hash cases separate formats.

Reply:
I see that this basically matches the [behavior](https://github.com/alpenlabs/zkaleido/blob/3ff1a395adf49876df4164494ea7d7e890602d5e/adapters/sp1/groth16-verifier/src/verifier.rs#L92-L95) of `zkaleido`. But the same question applies there!